### ISSUE TITLE
[sqlite] Database is locked for the CLI tools

### ISSUE THREAD

**Issue by mjnewton** [roundcube/roundcubemail#8035]
Environment:

Roundcube v1.4.11
Raspbian 10 (Debian Buster)
Apache 2.4.38
PHP 7.3
SQLite3 DB

Issue:

When attempting to use the moduserprefs.sh script to manipulate the database it sits there for ~60s and then reports that the database is locked (and the requested action isn't taken). For example:

> /moduserprefs.sh --user=13 mail_pagesize 150
Updating prefs for user 13...
ERROR: [5] database is locked (SQL Query: UPDATE "users" SET "preferences" = 'a:27:{s:6:"layout";s:4:"list";s:11:"search_mods";a:31:{s:1:"*";a:2:{s:7:"subject";i:1;s:4:"from";i:1;}s:4:"Sent";a:2:{s:2:"to";i:1;s:4:"body";i:1;}s:30:"NewtonNet etc.Logwatch Reports";a:3:{s:7:"subject";i:1;s:4:"from";i:1;s:4:"body";i:1;}s:5:"INBOX";a:3:{s:7:"subject";i:1;s:4:"from";i:1;s:4:"body";i:1;}s:7:"Archive";a:1:{s:4:"from";i:1;}s:26:"Archive.2013 Inbox Archive";a:3:{s:7:"subject";i:1;s:4:"from";i:1;s:4:"body";i:1;}s:26:"Archive.2011 Inbox Archive";a:2:{s:7:"subject";i:1;s:4:"from";i:1;}s:5:"Trash";a:3:{s:7:"subject";i:1;s:4:"from";i:1;s:4:"body";i:1;}s:26:"Archive.20
...[additional content snipped for brevity]... , "language" = 'en_GB' WHERE "user_id" = '13')
saved.

For what it's worth, this happens despite no instance of Roundcube running at the time (web service shut down) and, when it is up and running, all other aspects of Roundcube's functionaility (including reads and writes to the DB) are otherwise working fine.

It's been a very long time since I last used moduserprefs.sh and back then I was using a MySQL database so I don't know if that difference is significant.

Apologies if this isn't much to go on - I'm lacking experience with these particular aspects.

**Comment by mjnewton** [comment]
In case it helps, I've just tried with a brand new database with a single user in it (noting that as I say I used to use MySQL and converted that database to the sqlite3 one I use now which perhaps might've been a contributory factor?) and can confirm the same issue persists.

**Comment by alecpl** [comment]
In my opinion SQLite database should be used only for testing/development purposes, really.

We fixed a similar lock issue in installer by using transactions (#7064). Maybe it would help here too, but it's more like a hack. It would be good to find out why the database is locked in the first place.

**Comment by mjnewton** [comment]
> In my opinion SQLite database should be used only for testing/development purposes, really.

Ah okay. I actually selected it on the basis that this is a small system (~30 users) and so I figured it'd make more sense resources wise.

**Comment by mjnewton** [comment]
What feedback were you after @alecpl ? More than happy to provide it.

**Comment by alecpl** [comment]
Well, this is a label I put on tickets that I don't know what to do about. So, it's more like "need more info to decide what to do about it". Someone would have to try to reproduce the issue, and find out if it can be worked around with use of transactions. And of course someone could also investigate why the database is locked in the first place, but as I mentioned it might not be worth the effort.

**Comment by mjnewton** [comment]
I see - totally understand. I think I'll take your implied advice about not using sqlite for a production system. It's certainly an acceptable workaround for me.

**Comment by alecpl** [comment]
The issue is easily reproducible. Looks like the PHP process that created/used the file last leaves it in a locked state. 

Here's some findings:
1. If we add ` $dbh->setAttribute(PDO::ATTR_TIMEOUT, 5);` it will return the error in 5sec. instead of the default 60sec.
2. Using the database in [WAL](https://www.sqlite.org/wal.html) mode solves the issue, but I'm not sure about all the implications.

Here's the proposed patch:
```patch
--- a/program/lib/Roundcube/db/sqlite.php
+++ b/program/lib/Roundcube/db/sqlite.php
@@ -74,6 +74,12 @@ class rcube_db_sqlite extends rcube_db
                 }
             }
         }
+
+        // Enable WAL mode to fix locking issues like #8035.
+        $dbh->query("PRAGMA journal_mode=WAL");
+
+        // Change the default timeout (60) to a smaller value
+        $dbh->setAttribute(PDO::ATTR_TIMEOUT, 10);
     }
 
     /**
```

**Comment by alecpl** [comment]
Fixed.
